### ISSUE TITLE
Replies to posts can be indexed by search engines even if I've opted out of search engine indexing

### ISSUE THREAD

**Issue by jonpincus** [mastodon/mastodon#22047]
### Steps to reproduce the problem

1. Check the "opt-out of search engine indexing" option (Preferences/Other)
2. Make a public or unlisted reply to a public post by somebody who doesn't have that option checked
3. Wait for search engines to pick up that public post 
4. Use Google (or another search engine) to search for the content in your reply.  [Here's an example of a successful search using the text of an unlisted reply](https://www.google.com/search?q=%22one+gets+recognition+for+getting+a+lot+of+papers+and+also+for+attracting+a+lot+of+funds%22&client=ubuntu&hs=Dwm&channel=fs&sxsrf=ALiCzsb1tJVej1vsfJL-gGBhbVlueyP_iQ%3A1670136729744&ei=mUOMY-uHLZPA7gKBoI7gBQ&ved=0ahUKEwiruIKUsN_7AhUToFsKHQGQA1wQ4dUDCA8&uact=5&oq=%22one+gets+recognition+for+getting+a+lot+of+papers+and+also+for+attracting+a+lot+of+funds%22) -- thanks @dariusk !
...

### Expected behaviour

Google shouldn't return a link to a page with my post on it -- because I've opted-out of search engine indexing" should opt me out of search engine indexing

### Actual behaviour

Google returns the public post I replied to, with my reply on the page

### Detailed description

Checking the "opt-out of search engine indexing" should opt my posts out of indexing by search engines.  It doesn't.  

The description says that this option "Affects your public profile and post pages", and so people who are familiar enough with how federation and search engine's "noindex" directive works could figure out that this doesn't really opt you out. For most people, however, clicking the option just gives a false sense of security.

Either the behavior should be fixed to truly opt posts out of search engine results, or the option should be reworded to give a more accurate description -- "Include my profile in search engine results", for example.

### Specifications

Mastodon 4.0.2

**Comment by ineffyble** [comment]
This is unfortunately a difficult problem to solve, as there's no way to tell search engines to ignore *part* of a webpage.

**Comment by brad2014** [comment]
Would it be too crazy for UI to simply exclude from rendering reply toots on an indexable toot page for which context.descendants[x].author.noindex is true unless the user is authenticated?

**Comment by ineffyble** [comment]
It could make threads look very weird to remote/unauthenticated users, if random posts aren't rendered but replies to them are.

But another big problem is that the indexing preference is not federated, so even if you did this, it would only help where both the original post and the reply post are on the same instance.

**Comment by brad2014** [comment]
Understood.  Thank you for the explanation.

**Comment by jonpincus** [comment]
I didn't realize the "noindex" wasn't federated, that's certainly an issue.  Thanks for the info.

If it really can't be fixed then I recommend taking the option away (or maybe renaming it). The situation right now is the worst of all worlds: people *think* they're protected, but they aren't.

**Comment by skopf** [comment]
I would strongly suggest to completely remove this option as it gives a false sense of privacy.

Recent discussions in the Fediverse around search engines show impressively how this feature is misunderstood by users. This option is only setting the "noindex" meta tag on the profile page, but it does NOT prevent the profile from being indexed by other Mastodon instances.

Another problem is that this option has (almost) no impact on indexing by Google, although Google fully respects this option. Toots and profile information are not just displayed on the users profile page. If a post contains a tag, the post will also appear on the tag page. If a user has at least one follower on a remote instance, the toots will be federated there and be displayed on the federated timeline of the remote instance, from which Google will pick it up. Toots can also be boosted, or replied to, or...
Toots are simply displayed on too many pages to be able to hold up with the promise that they do not end up in Google.

I described this specific problem together with other privacy concerns here:
https://medium.com/@s.kopf/mastodon-is-a-zero-privacy-network-here-is-what-you-need-to-know-86aeed685e41

**Comment by jonpincus** [comment]
With #26344 federating a new opt-in setting for whether somebody wants their posts searchable, this is now fixable (using the new flag instead).

**Comment by ClearlyClaire** [comment]
> With #26344 federating a new opt-in setting for whether somebody wants their posts searchable, this is now fixable (using the new flag instead).

I haven't fully reviewed that PR yet, so I don't know what federates yet, but I have something to add here: having the information is only one part of the solution, the issue is that we can't tell search engines that some part of the page is not to be indexed. I have not found a way yet to exclude replies/parent posts from search results without also making them completely unavailable to logged-out users.

**Comment by brad2014** [comment]
Might we best let go of the idea that we present threads at all to unauthenticated users (which are in general a mix of *discoverable* and *not discoverable* posts), and simply present the one individual post for which the post URL refers, assuming that that one post is discoverable? 

That way the search result would give the searcher exactly what they asked for, and if they wanted to see more context, they could log in. 

The analogy for me is that of the embedded tweet - I see just the one (discoverable) tweet, not the entire thread.  If I click on it and log in (or am already logged in), I see the whole thread.

I'd be OK with that.

**Comment by brad2014** [comment]
On the other hand, an indexed toot with its thread that looks like:

------------
Toot text 1
------------
|
------------
Toot text 2
------------
|
-----------------------
Log in to read this toot
-----------------------
|
-----------------------
Toot text 4
-----------------------

Might have a great educational affect (and does not look too different from a content warning).

**Comment by jonpincus** [comment]
> having the information is only one part of the solution, the issue is that we can't tell search engines that some part of the page is not to be indexed. I have not found a way yet to exclude replies/parent posts from search results without also making them completely unavailable to logged-out users.

Agreed that search indexing is all-or-nothing.  One approach is to mark the page for any thread with a non-searchable reply as "noindex"; if replies aren't consenting to being searchable, the original poster's preference shouldn't override that.

**Comment by jonpincus** [comment]
> Furthermore, since Mastodon cannot ensure all replies federate everywhere and we are still lacking tools to alleviate that, a post's server remains the place where you are the most likely to have a complete view, and requiring you to register on every server would be very impractical.

Agreed.  By contrast, marking threads where the federated replies occur as "noindex" doesn't lead to these issues.  

Of course, badly-behaving search engines can still ignore noindex, so it's not an absolute protection, but the wording on the proposed privacy setting page highlighting that it doesn't _guarantee_ that posts are excluded covers this.
